### PR TITLE
Summary return types fixed, Server Tests added

### PR DESCRIPTION
**Pull request by Georgehe4**
Return type from AnalysisResult to a String wrapper class (since we cannot return as a String).
@lukechurch

### CODE REVIEW

**Comment by lukechurch** [comment]
@Georgehe4 This needs a test before we merge.

**Comment by Georgehe4** [comment]
Pushing this back until the week is done.

**Comment by Georgehe4** [comment]
@devoncarew @lukechurch 
Ready for review.

File: lib/src/dartpad_support_server.dart [hunk 1]
```diff
@@ -19,17 +20,75 @@ final Logger _logger = new Logger('dartpad_support_server');
...
+  var database;
+  bool test;
+
+  String getTypeName(dynamic obj) =>
+      mirrors.reflect(obj).type.reflectedType.toString();
+
+  String getClass(obj) =>
+      mirrors.MirrorSystem.getName(mirrors.reflectClass(obj).simpleName);
+
+  FileRelayServer({bool test}) {
```
Comment (devoncarew) [inline lib/src/dartpad_support_server.dart]: `bool test: false`

File: lib/src/dartpad_support_server.dart [hunk 2]
```diff
@@ -19,17 +20,75 @@ final Logger _logger = new Logger('dartpad_support_server');
...
     _logger.level = Level.ALL;
+    if (test) {
+      database = new Map();
+      this.test = true;
+    } else {
+      this.test = false;
+    }
+  }
+
+  Future<List> databaseQuery(var dbClass, String attribute, var value) async {
```
Comment (devoncarew) [inline lib/src/dartpad_support_server.dart]: If these methods are only used in this class lets make them private (here and `databaseCommit`)

File: lib/src/dartpad_support_server.dart [hunk 3]
```diff
@@ -19,17 +20,75 @@ final Logger _logger = new Logger('dartpad_support_server');
...
+  String getTypeName(dynamic obj) =>
+      mirrors.reflect(obj).type.reflectedType.toString();
+
+  String getClass(obj) =>
+      mirrors.MirrorSystem.getName(mirrors.reflectClass(obj).simpleName);
+
+  FileRelayServer({bool test}) {
     hierarchicalLoggingEnabled = true;
     _logger.level = Level.ALL;
+    if (test) {
```
Comment (devoncarew) [inline lib/src/dartpad_support_server.dart]: Using an `if` statements here totally works. Another way to do it is for the `ae.context.services.db.query` functionality to be injected in via the constructor. So, you have an interface which just exposes the operations you use. You have one implementation which implements that interface using the appengine class. It'll pretty much be a 1:1 passthrough for the calls you use. You then have a 2nd implementation which is a mock one. The real impl is the one that this class defaults to if none is provided.

File: test/dartpad_server_test.dart [hunk 4]
```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
```
Comment (devoncarew) [inline test/dartpad_server_test.dart]: 2015

**Comment by devoncarew** [comment]
I opened a bug for the test timeouts.

lgtm! @lukechurch?

File: lib/src/common_server.dart [hunk 5]
```diff
@@ -259,17 +264,17 @@ class CommonServer {
...
   }
   
-  Future<AnalysisResults> _summarize(String dart, String html, String css) async {
+  Future<SummaryText> _summarize(String dart, String html, String css) async {
     if (dart == null || html == null || css == null) {
       throw new BadRequestError('Missing core source parameter.');
     }
-    return _analyzeMulti({"main.dart" : dart}).then((result) {
-      Summarizer summer = new Summarizer(
+    return await _analyzeMulti({"main.dart" : dart}).then((result) {
```
Comment (lukechurch) [inline lib/src/common_server.dart]: I don't really like the return await, it's confusing as to the relationship between await, and then.

Let refactor this to the more standard await foo.
return bar;

File: lib/src/dartpad_support_server.dart [hunk 6]
```diff
@@ -59,16 +120,14 @@ class FileRelayServer {
...
     List result;
     do {
       randomUuid = new uuid_tools.Uuid().v4();
-      query = database.query(_GistMapping)..filter('internalId =', randomUuid);
-      result = await query.run().toList();
-      attemptCount ++;
-      if (!result.isEmpty) _logger.info("Collision in retrieving mapping id ${randomUuid}.");
+      result = await _databaseQuery(_GistMapping, 'internalId =', randomUuid);
+      attemptCount++;
+      if (!result.isEmpty) _logger
```
Comment (lukechurch) [inline lib/src/dartpad_support_server.dart]: This is awkward formatting please change to if (...) {
  _logger.info(...)
}

**Comment by lukechurch** [comment]
lgtm

**Comment by Georgehe4** [comment]
Merging.

**Pull request merged**